piDIOComm.c

```diff
@@ -151,70 +154,69 @@ INT32U piDIOComm_Init(INT8U i8uDevice_p)
 
 INT32U piDIOComm_sendCyclicTelegram(INT8U i8uDevice_p)
 {
-	INT32U i32uRv_l = 0;
-	SIOGeneric sRequest_l;
-	SIOGeneric sResponse_l;
-	INT8U len_l, data_out[18], i, p, data_in[70];
+	static INT8U last_out[40][DIO_OUTPUT_DATA_LEN];
+
+	INT8U i, p, data_in[70];
 	INT8U i8uAddress;
+	u8 in_buf[IOPROTOCOL_MAXDATA_LENGTH];
+	u8 out_buf[DIO_OUTPUT_DATA_LEN];
+	u8 snd_buf[DIO_OUTPUT_DATA_LEN];
+	u8 snd_len;
+	u8 rcv_len;
+	u8 cmd;
 	int ret;
-	static INT8U last_out[40][18];
-#ifdef DEBUG_DEVICE_DIO
-	static INT8U last_in[40][2];
-#endif
 
-	if (RevPiDevice_getDev(i8uDevice_p)->sId.i16uFBS_OutputLength != 18) {
+	if (RevPiDevice_getDev(i8uDevice_p)->sId.i16uFBS_OutputLength !=
+	    DIO_OUTPUT_DATA_LEN) {
 		return 4;
 	}
 
-	len_l = 18;
 	i8uAddress = RevPiDevice_getDev(i8uDevice_p)->i8uAddress;
 
 	if (piDev_g.stopIO == false) {
 		rt_mutex_lock(&piDev_g.lockPI);
-		memcpy(data_out, piDev_g.ai8uPI + RevPiDevice_getDev(i8uDevice_p)->i16uOutputOffset, len_l);
+		memcpy(out_buf,
+		       piDev_g.ai8uPI + RevPiDevice_getDev(i8uDevice_p)->i16uOutputOffset,
+		       DIO_OUTPUT_DATA_LEN);
 		rt_mutex_unlock(&piDev_g.lockPI);
 	} else {
-		memset(data_out, 0, len_l);
+		memset(out_buf, 0, sizeof(out_buf));
 	}
 
 	p = 255;
-	for (i = len_l; i > 0; i--) {
-		if (data_out[i - 1] != last_out[i8uAddress][i - 1]) {
+	for (i = DIO_OUTPUT_DATA_LEN; i > 0; i--) {
+		if (out_buf[i - 1] != last_out[i8uAddress][i - 1]) {
 			p = i - 1;
 			break;
 		}
 	}
 
-	sRequest_l.uHeader.sHeaderTyp1.bitAddress = i8uAddress;
-	sRequest_l.uHeader.sHeaderTyp1.bitIoHeaderType = 0;
-	sRequest_l.uHeader.sHeaderTyp1.bitReqResp = 0;
+	snd_len = DIO_OUTPUT_DATA_LEN;
 
 	if (p == 255 || p < 2) {
 		// nur die direkten output bits haben sich geändert -> SDioRequest
-		len_l = sizeof(INT16U);
-		sRequest_l.uHeader.sHeaderTyp1.bitLength = len_l;
-		sRequest_l.uHeader.sHeaderTyp1.bitCommand = IOP_TYP1_CMD_DATA;
-		memcpy(sRequest_l.ai8uData, data_out, len_l);
+		snd_len = sizeof(u16);
+		cmd = IOP_TYP1_CMD_DATA;
+		memcpy(snd_buf, out_buf, snd_len);
 	} else {
-		SDioPWMOutput *pReq = (SDioPWMOutput *) & sRequest_l;
+		SDioPWMOutput pwmOut;
 
-		memcpy(&pReq->i16uOutput, data_out, sizeof(INT16U));
+		memcpy(&pwmOut.i16uOutput, out_buf, sizeof(INT16U));
 
 		// kopiere die pwm werte die sich geändert haben
-		pReq->i16uChannels = 0;
+		pwmOut.i16uChannels = 0;
 		p = 0;
 		for (i = 0; i < 16; i++) {
-			if (last_out[i8uAddress][i + 2] != data_out[i + 2]) {
-				pReq->i16uChannels |= 1 << i;
-				pReq->ai8uValue[p++] = data_out[i + 2];
+			if (last_out[i8uAddress][i + 2] != out_buf[i + 2]) {
+				pwmOut.i16uChannels |= 1 << i;
+				pwmOut.ai8uValue[p++] = out_buf[i + 2];
 			}
 		}
-		len_l = p + 2 * sizeof(INT16U);
-		sRequest_l.uHeader.sHeaderTyp1.bitLength = len_l;
-		sRequest_l.uHeader.sHeaderTyp1.bitCommand = IOP_TYP1_CMD_DATA2;
+		snd_len = p + 4;
+		memcpy(snd_buf, &pwmOut.i16uOutput, snd_len);
```

Comment:
Hm, why can't we make `pwmOut` a pointer into `snd_buf` and thus directly operate on the data that's going to be sent? Similar to what's been done previously with `pReq`? That would avoid this newly introduced `memcpy()`. I'm worried that the `memcmp()` costs performance and this is a hotpath (executed every couple of msecs).

Reply:
The SDioPWMOutput type starts with a header, not with the data, so we cannot just cast snd_buf to this type, since snd_buf only contains the data. Thus the extra memcpy() (also I would not call this a "hotpath", since we do not want to be as fast as possible but only "fast enough"). However if we want to avoid the memcpy() and rather do the pointer conversion then we can define another struct for the pwm data (which does not have the header as a first element) and remove  SDioPWMOutput entirely, since it is not needed elsewhere.